### PR TITLE
Application Management API - Phase II

### PR DESCRIPTION
**Pull request by mefarazath**
This PR implements,

REST APIs for
- Creating Applications
- Updating an Application via PATCH
- Managing inbounds (GET, PUT, DELETE for each inbound)
- Regenerate Secret Key in OAuth inbound
-  Update Resident Service Provider via PATCH

Also includes,
- Improvements to exception handling (Sending proper error codes)
- Swagger improvements

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 1]
```diff
@@ -112,10 +112,10 @@ paths:
               description: Location of the newly created application.
               schema:
                 type: string
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/ApplicationModel'
+          # content:
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: why commented?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 2]
```diff
@@ -176,10 +182,10 @@ paths:
...
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: why commented?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 3]
```diff
@@ -285,33 +291,107 @@ paths:
       x-wso2-request: ''
       x-wso2-response: |
         "HTTP/1.1 200 Ok"
-    put:
+    # put:
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: why commented?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 4]
```diff
@@ -285,33 +291,107 @@ paths:
...
-            Location:
-              description: Location of the updated application
-              schema:
-                type: string
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/ApplicationModel'
+          description: Successfully Updated
+          # headers:
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: why commented?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java [hunk 5]
```diff
@@ -167,81 +218,310 @@ public ApplicationModel importApplication(InputStream fileInputStream, Attachmen
...
+            }
+            rollbackInbounds(getConfiguredInbounds(application));
+
+            String msg = "Error creating application.";
+            throw handleIdentityApplicationManagementException(e, msg);
         }
     }
 
-    public void updateApplication(String applicationId, ApplicationModel applicationModel) {
+    public void patchApplication(String applicationId,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java]: can't we inline line 255 and 256

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAuthenticationSequence.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        // iteration the options, divide in to federated and local and add the configs
+        if (CollectionUtils.isEmpty(stepModel.getOptions())) {
+            throw Utils.buildBadRequestError("Authentication Step options cannot be empty.");
+        }
+
+        List<LocalAuthenticatorConfig> localAuthOptions = new ArrayList<>();
+        List<IdentityProvider> federatedAuthOptions = new ArrayList<>();
+
+        stepModel.getOptions().forEach(option -> {
+            // TODO : add validations to swagger so that we don't need to check inputs here.
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateAuthenticationSequence.java]: shall we remove the todos?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/InboundAuthConfigToApiModel.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        return Collections.emptyList();
+    }
+
+    private InboundProtocolListItem buildInboundProtocolListItem(String applicationId,
+                                                                 InboundAuthenticationRequestConfig inbound) {
+
+        return new InboundProtocolListItem()
+                // TODO: add the friendly name from metadata service.
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/InboundAuthConfigToApiModel.java]: shall we remove todos

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml (hunk 3 above)
Comment (mefarazath) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: Cleaned up the commented swagger code

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/InboundAuthConfigToApiModel.java (hunk 7 above)
Comment (mefarazath) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/InboundAuthConfigToApiModel.java]: TODOs will be addressed in the next PR for application management REST APIs
Created https://github.com/wso2/product-is/issues/7000 to track all the TODOs

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/saml/ApiModelToSAMLSSOServiceProvider.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    private String getDefaultAssertionConsumerUrl(SAML2ServiceProvider spModel) {
+
+        if (spModel.getAssertionConsumerUrls().contains(spModel.getDefaultAssertionConsumerUrl())) {
+            return spModel.getDefaultAssertionConsumerUrl();
+        } else {
+            return spModel.getAssertionConsumerUrls().get(0);
+        }
+    }
+
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/saml/ApiModelToSAMLSSOServiceProvider.java]: unnecessary new line.

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 9]
```diff
@@ -1202,11 +1097,11 @@ paths:
             type: string
       responses:
         '200':
-          description: OK
-          content:
-            application/json:
-              schema:
-                $ref: '#/components/schemas/OpenIDConnectConfiguration'
+          description: Application Revoked
+          # content:
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: Remove commented part

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 10]
```diff
@@ -3567,17 +2610,18 @@ components:
             #     singleLogoutRequestUrl:
             #       type: string
             #       description: Single logout request accepting endpoint
-            #     singleLogoutMethod:
-            #       type: string
-            #       enum:
-            #         - backchannel
-            #         - frontchannel_http_redirect
+          #     singleLogoutMethod:
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: remove commented part

**Pull request merged**